packages/fela-bindings/src/FelaComponentFactory.js

```diff
@@ -1,35 +1,27 @@
 /* @flow */
-import resolveRule from './resolveRule'
-
 export default function FelaComponentFactory(
   createElement: Function,
   FelaTheme: Function,
   contextTypes?: Object
 ): Function {
   function FelaComponent(
-    { children, customClass, render = 'div', rule, style, ...restProps },
+    { children, as = 'div', style, ...otherProps },
     { renderer }
   ) {
-    return createElement(FelaTheme, {
-      render: theme => {
-        const props = rule ? { theme, ...restProps } : theme
-
-        const className = `${
-          customClass ? `${customClass} ` : ''
-        }${renderer._renderStyle(
-          resolveRule(rule || style, props, renderer),
-          props
-        )}`
+    return createElement(FelaTheme, undefined, theme => {
+      const className =
+        style instanceof Function
+          ? renderer.renderRule(style, { theme, ...otherProps })
+          : renderer._renderStyle(style, otherProps)
 
-        if (render instanceof Function) {
-          return render({
-            className,
-            theme,
-          })
-        }
+      if (children instanceof Function) {
+        return children({
+          className,
+          theme,
+        })
+      }
```

Comment:
When it's time, we have quite a large codebase that makes extensive use of both the HoC APIs and the render-prop api, so happy to help test codemods

Reply:
That'd be awesome!
One thing I stumbled over I'm not 100% sure how to convert is the following:

```js
const Button = createComponent(({ color }) => ({
  color,
  fontSize: '15px'
})

const ExtendedButton = createComponent(({ bgColor }) => ({
  backgroundColor: bgColor,
  fontSize: '18px',
  lineHeight: 1.2
}, Button)

// => color:red;background-color:blue;font-size:18px;line-height:1.2
<ExtendedButton color="red" bgColor="blue" />
```

How would you implement composition with `FelaComponent`? Of course you can explicitly accept e.g. extend or sth. but that seems like recuring boilerplate to me.
If we figure out user-friendly composition with render-props, I'm all in for replacing everything.

We can discuss that on Gitter if you want to!

Reply:
[react-adopt](https://github.com/pedronauck/react-adopt) seems to be the most popular solution for render-prop composition.

Right now, we use the `as` prop in a similar way to the second argument of `createComponent`, but only in the `string` for html primitive element case. What if we also allowed components there and automatically compose?

Reply:
Ok, I think I got it... 
We can actually pass style/rule to the render-props too.

```js
// Button.js
const button = ({ color }) => ({
  color,
  fontSize: '15px'
})

const Button = props => <FelaComponent style={button} {...props} />
```

```js
// ExtendedButton.js
import Button from './Button'

const extendedButton = ({ bgColor }) => ({
  backgroundColor: bgColor,
  fontSize: '18px',
  lineHeight: 1.2
})

const ExtendedButton = props => (
  <Button>
    {renderProps => <FelaComponent style={combineRules(renderProps.style, extendedButton)} />}
  </Button>
)

// => color:red;background-color:blue;font-size:18px;line-height:1.2
<ExtendedButton color="red" bgColor="blue" />
```

Reply:
And it is a bit harder to understand because none of the implementation detail is abstracted away. That's both good and bad, I guess.

Just thinking out loud here, but maybe if `style` is an array, its items can be automatically be passed to `combineRules`?

Reply:
Just realized - the `renderProps` object in the example above doesn't have a `style` property, only `className` and `theme`

Reply:
1. Yeah, I think we could expose the array-syntax with `combineRules` by default.
2. Yep the renderProps right now only has className and theme, we would have to expose that, but I got a new idea anyways:

```js
// ExtendedButton.js
import Button from './Button'

const extendedButton = ({ bgColor }) => ({
  backgroundColor: bgColor,
  fontSize: '18px',
  lineHeight: 1.2
})

const ExtendedButton = props => <Button extend={extendedButton} {...props} />

// => color:red;background-color:blue;font-size:18px;line-height:1.2
<ExtendedButton color="red" bgColor="blue" />
```

Reply:
Yes I like that. And unless I am missing something, it seems like it also solves the issue of adding lots of additional components to the React render tree.